### PR TITLE
SONARJAVA-1123 Add dot token to MemberSelectExpressionTree

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 1]
```diff
@@ -2107,14 +2113,17 @@ private ExpressionTree applySelectors(ExpressionTree primary, Optional<List<Expr
       for (ExpressionTree selector : selectors.get()) {
         if (selector.is(Kind.IDENTIFIER)) {
           IdentifierTreeImpl identifier = (IdentifierTreeImpl) selector;
-          result = new MemberSelectExpressionTreeImpl(result, identifier,
-            (AstNode) result, identifier);
+          // TODO The dot should have already been correctly handled
+          InternalSyntaxToken dotToken = InternalSyntaxToken.create(identifier.getFirstChild(JavaPunctuator.DOT));
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: At the moment I'm not able to find an way to get the `dot` more properly, as it has been introduced in the child of the `identifier`...

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 2]
```diff
@@ -2107,14 +2113,17 @@ private ExpressionTree applySelectors(ExpressionTree primary, Optional<List<Expr
...
+          result = new MemberSelectExpressionTreeImpl(result, dotToken, identifier,
+            (AstNode) result, dotToken, identifier);
         } else if (selector.is(Kind.METHOD_INVOCATION)) {
           MethodInvocationTreeImpl methodInvocation = (MethodInvocationTreeImpl) selector;
           IdentifierTreeImpl identifier = (IdentifierTreeImpl) methodInvocation.methodSelect();
-
-          MemberSelectExpressionTreeImpl memberSelect = new MemberSelectExpressionTreeImpl(result, identifier,
-            (AstNode) result, methodInvocation.getFirstChild(JavaPunctuator.DOT), identifier);
+          // TODO The dot should have already been correctly handled
+          InternalSyntaxToken dotToken = InternalSyntaxToken.create(methodInvocation.getFirstChild(JavaPunctuator.DOT));
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: At the moment I'm not able to find an way to get the `dot` more properly, as it has been introduced in the child of the `methodInvocationTree`...

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 3]
```diff
@@ -2107,14 +2113,17 @@ private ExpressionTree applySelectors(ExpressionTree primary, Optional<List<Expr
       for (ExpressionTree selector : selectors.get()) {
         if (selector.is(Kind.IDENTIFIER)) {
           IdentifierTreeImpl identifier = (IdentifierTreeImpl) selector;
-          result = new MemberSelectExpressionTreeImpl(result, identifier,
-            (AstNode) result, identifier);
+          // TODO The dot should have already been correctly handled
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 4]
```diff
@@ -2107,14 +2113,17 @@ private ExpressionTree applySelectors(ExpressionTree primary, Optional<List<Expr
...
+          InternalSyntaxToken dotToken = InternalSyntaxToken.create(identifier.getFirstChild(JavaPunctuator.DOT));
+          result = new MemberSelectExpressionTreeImpl(result, dotToken, identifier,
+            (AstNode) result, dotToken, identifier);
         } else if (selector.is(Kind.METHOD_INVOCATION)) {
           MethodInvocationTreeImpl methodInvocation = (MethodInvocationTreeImpl) selector;
           IdentifierTreeImpl identifier = (IdentifierTreeImpl) methodInvocation.methodSelect();
-
-          MemberSelectExpressionTreeImpl memberSelect = new MemberSelectExpressionTreeImpl(result, identifier,
-            (AstNode) result, methodInvocation.getFirstChild(JavaPunctuator.DOT), identifier);
+          // TODO The dot should have already been correctly handled
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 info

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java (hunk 3 above)
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: At the moment I'm not able to find an way to get the `dot` more properly, as it has been introduced in the child of the `identifierTree`...

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java (hunk 4 above)
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: At the moment I'm not able to find an way to get the `dot` more properly, as it has been introduced in the child of the `methodInvocationTree`...

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 5]
```diff
@@ -2074,60 +2079,65 @@ public ExpressionTree newIdentifierOrMethodInvocation(Optional<TypeArgumentListT
     return result;
   }
 
-  public ExpressionTree completeMemberSelectOrMethodSelector(AstNode dotTokenAstNode, ExpressionTree partial) {
-    ((JavaTree) partial).prependChildren(dotTokenAstNode);
-    return partial;
+  public InternalPrefixUnaryExpression completeMemberSelectOrMethodSelector(AstNode dotTokenAstNode, ExpressionTree partial) {
+    // Hack to be able to store the dot token and the epression
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: typo.

**Comment by m-g-sonar** [comment]
LGTM! You can probably squash the 2 last commits (3f07855 and c9c7d88)

**Pull request merged**